### ISSUE TITLE
Trait reform: big plans for nalgebra 0.11.0

### ISSUE THREAD

**Issue by sebcrozet** [sebcrozet/nalgebra#207]
---

This discussion is also on the [user forum](http://users.nphysics.org/t/trait-reform-big-plans-for-nalgebra-0-11-0/37).
**Please**, share what you would like to see added to the next release of nalgebra (in particular for computer-graphics related features)!

---

After #187, the next release will be the next big step toward what will be a version 1.0 in the future. It will include many breaking changes (especially for generic programming) but this is the last time it will break that much!

This trait reform is also a good opportunity to reorganize the internals of nalgebra, get rid of some hidden aberrations (like the ability to change the `Rotation2` diagonal to arbitrary values), and take in account negative feedbacks (I actually read IRC logs) regarding its ergonomy for computer graphics applications.

The main goals are to:
- allow generic programming through solid trait-based mathematical foundations. The current traits are too granular and not organized (with inheritance) in a coherent manner.
- use the [approx](http://crates.io/crates/approx) for approximate equality testing (i.e. don't redefine our own).
- add computer-graphics-related features to 3x3 and 4x4 raw matrices. The computer graphics community is used to work with 4x4 matrix to represent a general 3D transformation (or 3x3 matrices for 2D). Thus, people are usually frustrated when they can't do anything CG-related with `Matrix3` and `Matrix4`. This will change but they will still not be usable in a generic context (except for one case, see bellow).
# Generic programming

The new trait hierarchy will:
- mostly be implemented in a separate crate: [alga](https://github.com/sebcrozet/alga). Note that this is based on the excellent work of @brendanzab and @WaDelma on [algebra](https://github.com/brendanzab/algebra) (I wanted to give a fresh reboot of this library with different design choices).
- follow more strict mathematical notions and names. If several names are possible for the same structure, the simpler one is chosen (e.g. trait for rotations will be called `Rotation` instead of `SpecialOrthogonalGroup`).
- provide the same (or more) features that already exist on `nalgebra`. You'll just have to change the name of the traits you import, and sometimes the methods names as well. For example the `SquareMatrix`, `Diagonal`, `Column`, and `Row` will be merged into the traits `SquareMatrix` and `SquareMatrixMut` (the latter allows modifications).

Implementing your own instance of the traits of `alga` will be a bit harder as you'll have to
select the right stack of algebraic structures starting with the most basic one (the `Magma`). On this other hand, this will force you to be sure that your type has the correct algebraic properties that generic code will rely on. Convenient macros might be added in future versions of `alga` to make this easier.

Right after the release, a small guide will be provided to help you update your code without having to search too much what is moved under which trait.
# Computer Graphics features

It will be possible to:
- build a projection/look-at/rotation/translation/scaling `Matrix3/4` directy using dedicated
  constructors like `Matrix3::new_translation`.
- append/prepend a rotation/translation/scaling to `Matrix3/4`.
- transform a `Point2/3` or a `Vector2/3` using a `Matrix3/4`. The transformations will
  implicitly follow the rules of projective geometry (i.e. homogeneous coordinates). A
  `Transformation` trait will be the only one implemented by those matrices. Transformation will
  also work by simple multiplication.
- convert a `Rotation2/3`, `Isometry2/3` and `Similarity2/3` using the `na::cast` function (which will be renamed `na::convert`). Users won't have to use the `to_homogeneous` and `from_homegenous` if they are disturbed by the mathematical names. Under certain strict conditions checked at runtime, the inverse conversion (e.g. from `Matrix3` to `Rotation2`) will be possible (and will return `None` if it fails).
- multiply a `Matrix3/4` with a `Rotation/Isometry/Similarity`.

Most free functions at the root of the crate will be unchanged (except for the trait bounds of their arguments).
# Fixed issues

The following issues will be fixed: #205 , #200, #160, #145, #137, #36, and #32.

I have been working on this during the past couple of months and think I am halfway (not pushed yet).
Before the release, I'll also update kiss3d, ncollide, and nphysics in order to have a taste (and fix flaws) of the new design in extremely generic code.

**Comment by Andlon** [comment]
Not sure if this is desirable, but personally I find it very much so. Can I suggest perhaps considering the _removal_ of `Mul<VectorK<N>> for VectorK<N>` implementations, as well as the division operators? Perhaps instead replace them with functions that make it explicit that it is element-wise multiplication.

**Comment by sebcrozet** [comment]
@Andlon : yes, componentwise multiplication and division will become explicit methods only. It turns out that for this new design of nalgebra, vectors will just be the same thing as matrices so they will already have matrix-matrix-multiplication implemented (and componentwise multiplication operator would create conflicting impls).

To be more precise, we will deal with vectors kind of like Eigen does, i.e., Eigen has something like `using Vector<N, Dim> = Matrix<N, Dim, 1>` (where N is the scalar type and Dim the vector dimension). Similarly, nalgebra will have the type alias `type Vector<N, Dim> = Matrix<N, Dim, Static<U1>, S>` where `Static<U1>` means the number 1 known at compile-time, and `S` is the type of some data buffer that contains `Dim` elements. For example:
* `type Vector3<N> = Matrix<N, Static<U3>, Static<U1>, A>` is a 3D (column) vector.
* `type DVector<N> = Matrix<N, Dynamic, Static<U1>, B>` is a (column) vector with a size known at run-time.

Here, `A` and `B` are types that I did not write explicitly, but `A` is basically a statically-allocated array like `GenericArray` and the second one is dynamically allocated like a `Vec<N>`.

**Comment by paholg** [comment]
I have a rather niche request, which goes towards greater genericity. It makes some things rather cumbersome, though, and may cause some issues in edge cases, so I understand if it's off the table.

As far as I can tell, operators are only defined when the parametrized type is the same for both operands and is preserved under the operation. E.g. you define `Mul` between two vectors over `N` with the constraint `N:  Mul<N, Output = N>`.

I would like to see that change to `Mul` being between a vector over `M` and one over `N`, with the constraint simply `M: Mul<N>`, so that the result would be a vector over `<M as Mul<N>>::Output`.

I am not sure if there are other types that would benefit from this, but it would make `nalgebra` able to be parametrized over types from dimensioned.

Dimensioned has been in pretty much a proof-of-concept, but not really usable state for a long time, but I am currently rewriting it and it is very close to being actually usable.

**Comment by sebcrozet** [comment]
Before getting into Rust, I knew about a similar library called [dimensional](https://hackage.haskell.org/package/dimensional) in Haskell. I never used it but the idea. I think it is worth considering being compatible with `Dimensioned` but I'd prefer delaying that until after nalgebra 0.11.0 is released. This does not sound like a breaking change to `nalgebra` (this would only make things _more_ generic) and requires some non-trivial thinking which I'd prefer not doing now as I already have my hands full with the ongoing redesign.

My main concern here is that something like `Dimensioned` is not compatible with multiplicative abstract algebraic structures because it breaks the stability of multiplication (multiplying two `m.s⁻¹` gives you a `m².s⁻²` which is another type). But maybe this just means that it does not makes sense to used `Dimensioned` types with, say, matrices. It is compatible with additive structures like vector spaces though.

**Comment by paholg** [comment]
@sebcrozet That sounds very reasonable. I would be happy to help out with it when the time comes.

**Comment by sebcrozet** [comment]
Time for some updates on this! You will see the work done so far on the [nexgen](https://github.com/sebcrozet/nalgebra/tree/nexgen) branch. This is basically a full rewrite of the library internals with lots of added features (and a few removed).

Also, I'd appreciate some **help** from whoever would be interested in working on integrating Blas/Lapack/matrixmultiply. All the relevant code to be modified are in the `src/core/matrix.rs` and `src/core/ops.rs` files.

# Typenum-based dimension genericity
### (Almost) everything is generic wrt. the dimension.

For example a 3x3 matrix is `Matrix<f32, U3, U3, S>` while a (column) vector is `Matrix<f32, U3, U1, S>`. In both cases `S` is some other type parameters that represents the matrix data storage type (more on this later). Of course, many type aliases are defined for, e.g., low dimensional matrices and vectors. For example, `Vector1`  to `Vector6` are provided, as well as all square and rectangular matrices from `Matrix1` to `Matrix6` (for example a 2x3 matrix type alias is `Matrix2x3`).

### Dynamically-sized matrices and vectors
Dynamical-sized matrices and vectors are declared just like statically-sized matrices, but with a special value for its dimension, i.e., `Matrix<f32, Dynamic, Dynamic, S>` and `Matrix<f32, Dynamic, U1, S>` (for dynamically-sized vectors). This is actually must like Eigen. This is where the last type parameter `S` (which you don't have to worry about as long as you use the type aliases) plays its main role: dynamic entities will store their data as `Vec` while statically-sized matrices will use `GenericArray`.

Thus, because dynamic and static matrices use the same base type (`Matrix<...>`), they both share the same code.

### Vector/Matrix slicing
It is now possible to take references to parts of a vector or matrix. For example, it is possible to extract a 2x2 submatrix from a 3x3 matrix using:

```
let m3 = Matrix3::new(...);
let m2 = m3.fixed_slice::<U2, U2>(0, 0);
```

Here, `m2` will a 2D submatrix of `m3` starting with its top-left component. Because this slice is simply a `Matrix<f32, U2, U2, S>` with some special value for the `S` storage parameter (which contain an internal reference to `m3`), they share most capabilities of non-slice matrices (e.g.  matrix/matrix or matrix/vector multiplication, etc.)

Mutable slices, slices-of-slices, and custom strides are also possible.

If the slice size is not known at compile-time, it is possible to write, e.g.:

```
let m2 = m3.slice((0, 0), (2, 2));
```

Here, `m2` is a dynamically-sized slice and its type is `Matrix<f32, Dynamic, Dynamic, S>` with some special value of the `S` storage parameter which contain an internal reference to `m3`.

### Transformations
The whole set of common transformations are now included in nalgebra. In particular, we have types for:
  * Translation
  * Rotation matrix
  * Unit quaternion (3D rotation)
  * Unit complex (2D rotation)
  * Isometry (translation × rotation)
  * Similarity (translation × rotation × uniform scale)
  * Affine (translation × rotation × non-uniform scale × rotation)
  * Projective (arbitrary invertible transformation).

They can be used in generic code as they implement transformations traits from [alga](https://github.com/sebcrozet/alga).


### Less traits
If you don't care about generic programming, you should not have to import tons of traits to use algebraic operators. That's why most operators are now implemented as struct methods. Use traits from **alga** for generic programming.

# What's next
A few feature still have to be added or polished. Then, the next steps are:

### Documentation
I'm working on a new website (much like the one for [ncollide](http://ncollide.org)) that will document all those features. The rustdoc-generated doc put too much emphasis on trait bounds, making it very hard to read.

### Performance
We basically need to integrate [matrixmultiply](https://github.com/bluss/matrixmultiply), Blas and Lapack. **Any contribution on that would be strongly appreciated.** We also need to check that we did not regress from the previous version of nalgebra.

# Unresolved issues
I get a *lot* of `Broken MIR` errors when compiling tests of benchmarks (mostly because it fails to recognize that an associated type is equal to some other exact type). I am not sure how to decrease the code base to report this as an issue to the rust compiler repo.

**Comment by sebcrozet** [comment]
This is done. The `Broken MIR` errors are still there and blas/lapack/matrixmultiply are not yet integrated. I'm creating new issues for those.
